### DISCUSSION TITLE
MET V10.0 Series-Analysis NetCDF graphical output

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#1434] [Plot Generation] [answered]
When I use a masking file to perform verification of model output for a larger domain over a smaller, included domain nested 
inside the larger domain, the resulting graphical output (MET V10.0 Series-Analysis NetCDF file) produced by MET Plot-Data-Plane 
shows the area included within the boundaries of the larger domain surrounding the smaller masking region within which the 
verification is performed. I want to eventually publish these graphics along with other graphics where the verification was 
performed without a masking file, but the differences in the graphics make comparison more difficult. Should I be running 
Series-Analysis differently to produce graphical output which displays just the verification masking region only? I checked 
the MET Plot-Data-Plane options and I don't see any options to control the plotting area, but maybe I missed something. When 
I use ncview, I get the same issue. Thanks.
John Raby

A related follow-up question: Is it possible get Plot-Data-Plane to generate a plot of the NetCDF output from Series-Analysis over a user-specified grid?

**Comment by jprestop** [comment]
Hi @jwraby.  

I needed to get @JohnHalleyGotway's expertise on this one.  He was able to provide a lot of great information.

Series-Analysis writes a NetCDF output file containing gridded data. The dimensions of the data written to that file are the same as the dimensions of the verification grid. There is no mechanism in MET by which to subset the data written to that output file.
However, there are some relevant details to consider...

Instead of restricting the processing in Series-Analysis by defining a polyline, you could regrid the input data using the "[regrid](https://github.com/dtcenter/MET/blob/9c45689439be6b1eaff6fd47b320cb5a511b05e6/met/data/config/SeriesAnalysisConfig_default#L30)" dictionary of the config file. And by regrid, it's really subsetting the grid. If this is a simple lat/lon projection, you could define a subset of that grid using a [grid specification string](https://met.readthedocs.io/en/develop/Users_Guide/appendixB.html?highlight=grid%20specification#grid-specification-strings) using nearest neighbor interpolation.
```
regrid = {
   to_grid    = "GRID SPEC STRING HERE";
   method     = NEAREST;
   width      = 1;
   vld_thresh = 0.5;
   shape      = SQUARE;
}
```

Next, you're welcome to use the output of plot_data_plane however you'd like, but it isn't intended as a fully featured plotting tool.
That being said, there is related functionality in MODE. The [plot_valid_flag](https://github.com/dtcenter/MET/blob/9c45689439be6b1eaff6fd47b320cb5a511b05e6/met/data/config/MODEConfig_default#L217) instructs MODE to zoom the plot to only contain the region of valid data. That logic could be ported over to plot_data_plane and plot_point_obs. But that would require a new development issue.

I hope this is helpful information.

**Reply by jwraby** [reply]
Thanks for coordinating this with John. This information is helpful and increases my understanding of the situation we are dealing with. Is it possible/feasible for one to edit the NetCDF file in such a way as to remove the missing data leaving only the valid verification results in the inner domain and end up with a zoomed in plot? I'm thinking of another approach, perhaps a tool in nco which might do this.

**Reply by jwraby** [reply]
If I am already using the regrid dictionary of the Series-Analysis config file to perform the needed regridding of the gridded observations to the grid of the forecast grid, how can I also use the same dictionary to accomplish the "regridding" to a grid defined by the grid specifications? Is it possible to use two sets of settings in the regrid dictionary to first accomplish the regridding of the observations and then accomplish the subsetting of the grid?

**Reply by jprestop** [reply]
Hi @jwraby.  I'm going to ask @JohnHalleyGotway to jump in here and see if he can provide the answers to your questions.

**Reply by jwraby** [reply]
Referring to grid specifications for a lambert conformal grid, I looked at a metadata view of the grid I want to use but it doesn't list a lon_orient. It does show something called "central meridian". Is this the same spec?

**Reply by JohnHalleyGotway** [reply]
@jwraby if you're working with a Lambert Conformal projection, then defining a subset is NOT easy, like it would be with a lat/lon projection. Seems to me that you'd really like us to enhance plot_data_plane to support that plot_valid_flag option from MODE. I think that's really the functionality you're really after... and it does seem pretty reasonable.

To answer your question about the regrid dictionary, you're right, you only get to choose one "to_grid" setting. Sounds like you're probably setting "to_grid = FCST" to regrid the observations to the forecast domain. Here's the documentation for the [regrid dictionary](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html#regrid). Note that "to_grid" can be set several ways. In this case, you *could* define a Lambert Conformal [grid spec string](https://met.readthedocs.io/en/develop/Users_Guide/appendixB.html#grid-specification-strings) to regrid both the forecast and observation data to some smaller domain. You'd define that smaller grid to include the same spatial area your polyline currently does.

But I realize that isn't optimal since you'd be changing the (lat,lon) grid points of the forecast domain. I'll take a look at the plot_valid_flag for plot_data_plane.

**Reply by jwraby** [reply]
Thanks for looking into this, John. I like your suggestion about enhancing plot_data_plane to allow a fix for this issue. You are correct about the to_grid = FCST. I have a GRIB file (post-processed WRF output) which is for a domain identical to that of the polyline file, so if I interrogate the GRIB file for the lambert specs, won't I have the information needed to enable the regridding to produce the subset as you suggested earlier? If I use the one available "to_grid" setting to provide the grid specs, will that effectively regrid the observations as well as the fcst to the domain defined by the grid specs?

**Reply by JohnHalleyGotway** [reply]
John, it's even easier than that. Look through the regrid.to_grid options:

https://met.readthedocs.io/en/develop/Users_Guide/config_options.html?highlight=Regrid#regrid

You can set to_grid = path to a gridded data file. So just provide the path to the GRIB file that's already on the grid you'd like to use.

Does that do it?

**Reply by jwraby** [reply]
Very good point, John! I will test it and let you know the result.

**Reply by jwraby** [reply]
One point about your suggested approach is that instead of regridding the input gridded obs from 2.5 km grid spacing to the 3 km grid spacing of the input WRF forecast (to_grid = FCST), both the input forecast and gridded obs will be each be regridded to the 1 km grid of the target gridded data file. While this may achieve the desired reduction of the valid data area to the innermost nest and fix the plotting issue, I think it also changes the character of the verification some. My intent was to verify the 3 km WRF over the 1 km WRF domain (inner domain centered inside the larger 3 km WRF domain) in order to compare performance of both models over the same common domain. One benefit of doing it this way is that the comparison will be conducted all on a 1 km grid, but I'm wondering about the validity of this approach because of the regridding of the input forecast data from it's original 3 km grid spacing to a 1 km spacing before the verification is done. What is your opinion about this?

**Reply by JohnHalleyGotway** [reply]
@jwraby I do agree that regridding data from a coarser scale (3 km) to a finer scale (1 km) is generally not recommended. As I understood it, your motivation for doing this is to easily produce graphics for a paper.  Relatively speaking, I would be more concerned about doing this in Grid-Stat and less concerned about it in Series-Analysis. Introducing "extra" points in the computation of bulk statistics in Grid-Stat would yield an artificially inflated level of confidence in the resulting confidence intervals... that's really an artifact of regridding. In Series-Analysis, data for each grid point is processed independently. The output shows the geographic distribution of performance. The large-scale features likely matter more than the small-scale ones. And the number of grid points doesn't affect the CI's computed by Series-Analysis, only the length of the series does.

I don't have any obvious answers for you. You have several options:
- Verify both on their native resolutions (1km and 3km) and compare stats.
- Compare Fractions Skill Score from a 3x3 neighborhood of the 1km grid to stats from 3km output. 
- Verify both on the 1km grid.
- Verify both on the 3km grid.
- Interpolate both to the grid of analysis/observation data, assuming that differs from the model output.

Ultimately, it's really up to you and the reviewers of your paper. If ease in plotting is a complication, you could run Series-Analysis at the 3km resolution and use some other method to plot the NetCDF output.

**Reply by jwraby** [reply]
John - Thanks for providing your feedback on my comment. You are correct that the original issue was to run Series-Analysis in a way to produce graphics that would facilitate a particular comparison which was needed for publishing the results. The graphics produced using your suggestion will definitely work for us, but I will need to provide some information to the readers on the impact on the results of the processing method in order to produce those graphics. Fortunately, we also have the results from previous Series-Analysis runs where each WRF model (1 km and 3km ) was verified over it's own domain independently which is one of the options you listed above. I appreciate hearing about the other options for performing the verification. Regarding your comment about the particular way Series-Analysis works which may mitigate the potential issues associated with regridding, having the information on how the CI's are not affected by the number of grid point and that the length of the series does affect the CI's is very useful. Our input dataset includes 24 hourly input files over a 99-day period, so hopefully, the CI's will be narrowed down some.

**Reply by jwraby** [reply]
Just an update on the effect of regridding the fcst and obs input to a 1km grid prior to verification. I compared the Series-Analysis graphic for the 1km WRF over its own domain with the graphic of the 3km WRF (and 2.5 km gridded obs) both regridded to a 1km grid using the to_grid = path to a GRIB file generated by post-processing the output of the 1km WRF. One noticeable difference is that the granularity of the pixels in the former shows much finer detail than the latter. I can see larger pixels in the graphic which was 3km WRF and 2.5 km gridded obs regridded to a 1km grid prior to verification. My interpretation of this is that, even though the output verification grid is 1km, the data originated from lower resolution fields from the input 3km WRF and the 2.5 km gridded obs, so the depiction of the scores information can only be as detailed as the original input data. Does this make sense?

**Reply by JohnHalleyGotway** [reply]
Yes, that does make sense. Of course the result of the regridding depends a lot on the regridding method you choose. Selecting nearest neighbor will run very fast but the output will look like it has about the same resolution as the input. Selecting bilinear interpolation will run slower but the output will look smoother than the input because its interpolating values to locations between the original grid points.

**Reply by jwraby** [reply]
John - Thanks for sharing that info. We used the nearest neighbor method, so your description indicating the tendency for that method to retain the resolution of the original input agrees with what I'm seeing. Good point about the impact of the interpolation method. Perhaps using the nearest neighbor method mitigates to a certain extent the issues you mentioned earlier regarding regridding from coarser to finer.

**Comment by JohnHalleyGotway** [comment, marked as answer]
OK, proof of concept here. I made some changes to plot_data_plane locally and was able to get it to "zoom up" to the region of valid data. Original on the left and zoomed on the right.

The remaining tasks would be tidying up the code a bit, deciding how to make this configurable in plot_data_plane and plot_point_obs, updating the docs, and adding a test for each.

@TaraJensen should I write this up as an issue and include it in met-10.1.0?
<img width="1104" alt="Screen Shot 2022-02-11 at 6 09 43 PM" src="https://user-images.githubusercontent.com/21087144/153690226-c21f95ef-43bf-495d-b947-f55f08e4a18d.png">

**Reply by jwraby** [reply]
Thanks for working on this change which should address the issue nicely. I am having trouble downloading the image by just clicking on it. I will continue attempts to download, but it may be a firewall issue from an ARL VPN which is blocking. If I can assist in testing let me know.

**Reply by JohnHalleyGotway** [reply]
John, I wrote up this feature request in this new GitHub issue:
https://github.com/dtcenter/MET/issues/2052

Unfortunately, due to tight timelines, I doubt we'll be able to include in MET version 10.1.0. So I've assigned it to the MET version 11.0.0 release. Once you confirm that setting regrid.to_grid = "path to GRIB file" solves your immediate issue, we can probably mark this discussions as being "answered".

**Reply by jwraby** [reply]
Thanks for creating the enhancement request. I will let you know the results.

**Reply by jwraby** [reply]
The test, conducted earlier today, was successful. The output graphic produced by plot_data-plane is zoomed in to the inner domain where the valid verification data is. The test was conducted with the masking file defining the inner domain boundaries, but this was probably not necessary since the regrid.to_grid = "path to GRIB file" should ensure that the verification will occur only over the domain of the GRIB file. Your assistance in resolving this issue is greatly appreciated! I would be interested in any feedback you may have on my earlier comment about the validity of using this approach as opposed to the previous approach.

**Comment by JohnHalleyGotway** [comment]
@jwraby can you please select one of the responses in this discussion as being a good "answer". Just look for the **Mark as answer** button. Once you've picked a good answer, I'll **Lock conversation** on this discussion to prevent future posts. That's basically how we resolve Q&A's through GitHub discussions. Doing so makes it easier for the support team to figure out which discussions require attention.

Of course, if more questions arise, you can always start a new discussion.
